### let's try hand-coding the CFR algorithm for Kuhn Poker in openspiel

In [42]:
import pyspiel
import numpy as np
from collections import defaultdict

In [2]:
game = pyspiel.load_game("kuhn_poker")

In [87]:
from typing import Dict

cumulative_regrets = [defaultdict(lambda: defaultdict(float)), defaultdict(lambda: defaultdict(float))]
# cumulative_regrets[P][I][A] is the cumulative regret, for player P (0 or 1) at infostate I, of not taking action A
strategy_sums = [defaultdict(lambda: defaultdict(float)), defaultdict(lambda: defaultdict(float))]

InfoSet = str

def get_strategy_from_regrets(regrets: Dict[InfoSet, float]):
    positive_regrets = {action: max(value, 0) for action, value in regrets.items()}
    denom = sum(positive_regrets.values())
    if denom <= 0:
        return {action: 1/len(regrets) for action in regrets}
    return {action: value/denom for action, value in positive_regrets.items()}

"""
returns the expected value of the state for player 1 (not counterfactual, so the caller should do the scaling)
p1prob is the probability that we reach this state if *p2* plays to reach the state.
"""
def cfr(state, player, p1prob, p2prob):
    if state.is_terminal():
        return state.returns()[0] # take the 0th item here bc that's the payoff for player 1
    if state.is_chance_node():
        expected_value = 0
        for action, p in state.chance_outcomes():
            ev = cfr(state.child(action), player, p1prob*p, p2prob*p)
            expected_value += p*ev
        return expected_value
    else:
        infostate_str = state.information_state_string()
        legal_actions = state.legal_actions()
        # hacky way to lazily initialize cumulative regrets:
        for action in legal_actions:
            cumulative_regrets[state.current_player()][infostate_str][action] += 0
        
        strategy = get_strategy_from_regrets(cumulative_regrets[state.current_player()][infostate_str])
        # the following two variables store ev for the state.current_player().
        expected_value = 0
        expected_values_per_action = dict()
        for action in legal_actions:
            p = strategy[action]
            if state.current_player() == 0:
                ev = cfr(state.child(action), player, p1prob*p, p2prob)
            else:
                ev = cfr(state.child(action), player, p1prob, p2prob*p) * -1
            expected_values_per_action[action] = ev
            expected_value += ev * p
        if state.current_player() == player:
            if player == 0:
                pi_not_i, pi_i = p2prob, p1prob
            else:
                pi_not_i, pi_i = p1prob, p2prob
            for a, v in expected_values_per_action.items():
                counterfactual_regret = (v - expected_value)*pi_not_i
                strategy_sums[player][state.information_state_string()][a] += p1prob*p2prob * strategy[a]
                cumulative_regrets[player][state.information_state_string()][a] += counterfactual_regret
        # since expected_value stores the ev for the state.current_player(), we invert it before returning
        if state.current_player() == 0:
            return expected_value
        else:
            return expected_value * -1
        
    

In [100]:
ITS = 6000
cumulative_regrets = [defaultdict(lambda: defaultdict(float)), defaultdict(lambda: defaultdict(float))]
strategy_sums = [defaultdict(lambda: defaultdict(float)), defaultdict(lambda: defaultdict(float))]

for it in range(ITS):
    for player in [0,1]:
        state = game.new_initial_state()
        payoff = cfr(state, player, 1, 1)
        if it%100 == 0:
            print(it, payoff)

0 0.3125
0 0.25
100 -0.06588413371255253
100 -0.048032432641927136
200 -0.06906437889014388
200 -0.05355714547129836
300 -0.06566306853256654
300 -0.054341641765788706
400 -0.06513692336808857
400 -0.054513462324123074
500 -0.062502406696407
500 -0.0514975003093206
600 -0.06048811054526626
600 -0.05187123758438289
700 -0.06165555781247328
700 -0.05274160977037906
800 -0.06047190947007608
800 -0.052812115685123284
900 -0.05991325207071202
900 -0.05264948992724433
1000 -0.06118483326617913
1000 -0.05399961160211475
1100 -0.060677978013210065
1100 -0.05381614968913967
1200 -0.06039161368689444
1200 -0.05383466878256843
1300 -0.0597133513026481
1300 -0.053350190832990974
1400 -0.06009304855012565
1400 -0.05392673665614339
1500 -0.05941286367004783
1500 -0.053638321238782205
1600 -0.059533356423190964
1600 -0.05382086273486136
1700 -0.05989809797001355
1700 -0.054242198132835484
1800 -0.05946306037069471
1800 -0.05413928180507116
1900 -0.05905168187522064
1900 -0.053818286059890974
2000 -0.

In [94]:
def get_strategy_from_strategy_sums(strategy_sums):
    denom = sum(strategy_sums.values())
    if denom <= 0:
        return {action: 1/len(strategy_sums) for action in strategy_sums}
    return {action: p/denom for action, p in strategy_sums.items()}
    

def get_payoff_with_final_strat(state):
    if state.is_terminal():
        return state.returns()[0]
    elif state.is_chance_node():
        ev = 0
        for action, p in state.chance_outcomes():
            ev += p*get_payoff_with_final_strat(state.child(action))
        return ev
    else:
        infostate_str = state.information_state_string()
        legal_actions = state.legal_actions()
        print(' '*state.move_number(), state.history_str())
        strategy = get_strategy_from_strategy_sums(strategy_sums[state.current_player()][infostate_str])
        ev = 0
        for action in legal_actions:
            print(' '*state.move_number(),infostate_str,'->',state.action_to_string(action), strategy[action])
            ev += strategy[action] * get_payoff_with_final_strat(state.child(action))
        return ev

In [101]:
state = game.new_initial_state()
ev = get_payoff_with_final_strat(state)
print(ev)

   0 1
   0 -> Pass 0.764724189393865
    0 1 0
    1p -> Pass 0.9996989973206932
    1p -> Bet 0.00030100267930670705
     0 1 0 1
     0pb -> Pass 0.9999725586952606
     0pb -> Bet 2.7441304739366313e-05
   0 -> Bet 0.23527581060613498
    0 1 1
    1b -> Pass 0.6561151815826567
    1b -> Bet 0.3438848184173433
   0 2
   0 -> Pass 0.764724189393865
    0 2 0
    2p -> Pass 0.0001196960864903472
    2p -> Bet 0.9998803039135097
     0 2 0 1
     0pb -> Pass 0.9999725586952606
     0pb -> Bet 2.7441304739366313e-05
   0 -> Bet 0.23527581060613498
    0 2 1
    2b -> Pass 0.0003615765470801664
    2b -> Bet 0.9996384234529199
   1 0
   1 -> Pass 0.9988066610907277
    1 0 0
    0p -> Pass 0.66381571464462
    0p -> Bet 0.33618428535538003
     1 0 0 1
     1pb -> Pass 0.4286631829955627
     1pb -> Bet 0.5713368170044373
   1 -> Bet 0.0011933389092724158
    1 0 1
    0b -> Pass 0.9998798988746334
    0b -> Bet 0.00012010112536670338
   1 2
   1 -> Pass 0.9988066610907277
    1 2 0
   

The resulting EV looks correct (https://en.wikipedia.org/wiki/Kuhn_poker): "the first player should expect to lose at a rate of −1/18 per hand"

In [96]:
1/18

0.05555555555555555

Ok that looks correct.

Player one should bet 3x more likely with a K (2) than with a J (0): https://upload.wikimedia.org/wikipedia/commons/a/a9/Kuhn_poker_tree.svg

```
0 -> Bet 0.23675878818473314
...
2 -> Bet 0.6698347958294653
```

In [98]:
0.237 * 3

0.711

Hm.... that doesn't look exactly correct.  [edit: Re-ran with 5x more iterations (6000 iterations) and got closer values (0.235 and 0.688), so I'm satisfied]

More checking:  
Player one should check 100% of the time with a Q (1), which it does: `1 -> Pass 0.9928399665443669`  
Player two should fold 100% of the time with a J when bet to, which it does: `0b -> Pass 0.9992716750496424`  
And player two should bet 33% of the time with a J when checked to, which it does: `0p -> Pass 0.6608318109822778`

Alright now let's try using openspiel's CFR solver: [cfr_example.py](https://github.com/deepmind/open_spiel/blob/master/open_spiel/python/examples/cfr_example.py)


In [105]:
from open_spiel.python.algorithms import cfr
from open_spiel.python.algorithms import exploitability

cfr_solver = cfr.CFRSolver(game)
ITS = 1000
for i in range(ITS):
    cfr_solver.evaluate_and_update_policy()
    if i % 100 == 0:
        conv = exploitability.exploitability(game, cfr_solver.average_policy())
        print("Iteration {} exploitability {}".format(i, conv))

Iteration 0 exploitability 0.45833333333333326
Iteration 100 exploitability 0.007681433241925523
Iteration 200 exploitability 0.00344489604438572
Iteration 300 exploitability 0.0029364753880415373
Iteration 400 exploitability 0.0018367557313718752
Iteration 500 exploitability 0.0012031926490508604
Iteration 600 exploitability 0.0013515037961346898
Iteration 700 exploitability 0.0009143638829439416
Iteration 800 exploitability 0.0008587977290051563
Iteration 900 exploitability 0.001075271558795332


that was easy.  Let's try to compare the exploitability of the resulting policy from my hand-coded CFR.

In [114]:
from open_spiel.python.policy import TabularPolicy
my_policy = TabularPolicy(game)
for player in [0,1]:
    for info_state in strategy_sums[player]:
        for action, p in strategy_sums[player][info_state].items():
#             print(info_state, action, p)
            my_policy.policy_for_key(info_state)[action] = p

0pb 0 126.52922202103277
0pb 1 0.003472222222222222
0 0 254.9080631312879
0 1 78.42527020204486
1pb 0 95.18955823728828
1pb 1 126.87186904015913
1 0 332.9355536968497
1 1 0.3977796364240675
2pb 0 0.003472222222222222
2pb 1 17.09382441962857
2 0 103.99172242819718
2 1 229.34161090513626
1p 0 179.53995030931154
1p 1 0.05405827777214395
1b 0 100.8707049722805
1b 1 52.86861977396898
2p 0 0.03530092592592592
2p 1 294.8860031951803
2b 0 0.013888888888888888
2b 1 38.3981403233087
0p 0 144.50607379338578
0p 1 73.18397271409782
0b 0 115.62939793695978
0b 1 0.013888888888888888


In [112]:
conv = exploitability.exploitability(game, my_policy)
print("my exploitability: {}".format(conv))

my exploitability: 14728.708476786192
